### ISSUE TITLE
Handle wide documents without widening all paragraphs

### ISSUE THREAD

**Issue by sgtatham** [jugglerchris/rust-html2text#98]
### Motivation

If the input to `html2text` can't be rendered in the specified display width, one option (in some client UIs) is to render it anyway at a larger width, and do some other kind of compromise when presenting it to the user, such as apologetically telling them they need to widen their terminal, or providing keystrokes to pan the window left and right across the wider logical canvas that you rendered the document on.

In the current API, you can do this by handling `Error::TooNarrow`: if you try to render at the physical terminal width and get `TooNarrow`, you can increase the render width and try again, until you find a width at which the document successfully renders. (And then maybe you binary-search between the largest failing width and the smallest successful one to find the cutoff? And then maybe you increase it by 20% or 50% or something, to avoid one-character-wide table cells? But these are client-side decisions.)

But the downside of doing this is that if the document contains one table that needs at least 500 columns, then all the ordinary paragraphs alongside that table are _also_ wrapped at 500 columns, making them very hard to read. It would be nicer if all the parts that don't _have_ to be ultra-wide could still be made to fit into my (say) 80-column terminal.

In particular, if the wide table is something that not all readers of the document will need to bother with at all, then that way, I can read the introductory paragraphs easily, and use them to decide whether I need to bother widening my window to read the table!

### Suggestions

There's more than one way that that high-level requirement might be achieved. Two thoughts that have sprung to mind (which may still not be the only possible approaches):

1\. Client application tells `html2text` to render at my actual terminal width, but with a flag of some kind that says that that width limit is 'strongly advised' rather than 'mandatory'. Then if a table can't help being wider than specified, fine, it can do that. But anything that _can_ fit in the width should.

After setting this flag, the client would never expect to get `Error::TooNarrow` back from the renderer. But, in return, it accepts that the output might contain overlong lines, and must find some way of dealing with them itself if so.

2\. Client application gives `html2text` _two_ width parameters. One is the physical display width, just as now. The other is a 'maximum wrap width'. Each paragraph is wrapped to min(maximum wrap width, however much space is physically available in the layout).

The idea is that the client starts by setting both widths to the physical terminal width. If the document can't be rendered at that width, then the renderer still returns `Error::TooNarrow`, and the client responds the same way I suggested above, by cranking up the display width until the error stops happening. But it leaves the maximum wrap width at its original value.

### Pros and cons

Option 1 has the advantage of only needing one call to the renderer, instead of O(log _n_) calls to zero in on a good render width.

Option 2 gives the client more control, because it still gets to decide how much _larger_ than the absolute minimum possible width it wants to make things.

### Fiddly edge cases

In both cases, there's a question of what happens to a single very wide table cell. (Say, one row of your table has an enormous number of tiny cells, and another has a single cell with `colspan=lots` covering the same horizontal space as all of them, so that that one cell is forced to be wider than the physical display.) With my option 2, the very wide cell would still have its paragraphs limited to the max wrap width, because that max width is applied to every paragraph anywhere in the whole layout. So if the client application is handling wide documents by providing keystrokes to pan left and right across the canvas, there will be _some_ pan position at which the text of that table cell is readable. Perhaps option 1 might do the same thing?

With the max-wrap-width approach, there's also a question of how you measure it for indented things like `<li>` or `<blockquote>`: from the left margin of the _text_, or of the containing column? For example, which of these would you see, in a set of nested lists? This, which lets you read the whole list without having to pan your display at all?

```
plain para at the
full screen width

* narrower bullet
  point

  * even narrower
    bullet point

    * smaller yet
      and so on
```

Or this, which avoids the bullet points getting squashed up against the right-hand edge, making use of the fact that they physically _do_ have space to expand into?

```
plain para at the
full screen width

* bullet point uses
  same width so its
  margin is 2 chars
  further right

  * nested bullets in
    turn move 2 chars
    right each time

    * result: you never
      get text squashed
      too narrow
```

The CSS `max-width` property takes the latter view, on the basis that it expects that you've widened your browser window and the whole wide window is visible. But in a context where you might be using a "pan left and right" UI, perhaps the former makes more sense?

Or perhaps some further compromise involving a _minimum_ wrap width, so that once the bullet points get absolutely too silly squashed up against the right margin, there's a fallback available?

As you can see, I don't have all the answers here :-)

**Comment by jugglerchris** [comment]
I have implemented the second option, which is available in 0.9.0, as that was straightforward, as [`Config::max_wrap_width()`](https://docs.rs/html2text/latest/html2text/config/struct.Config.html#method.max_wrap_width).

[`Example usage`](https://github.com/jugglerchris/rust-html2text/blob/main/src/tests.rs#L19) would be something like:

```rust
    let result = config::plain()
        .max_wrap_width(wrap_width)
        .string_from_read(input, width)?;
```

**Comment by jugglerchris** [comment]
Option 1 has also now been implemented in 0.12.0, with [`allow_width_overflow()`](https://docs.rs/html2text/latest/html2text/config/struct.Config.html#method.allow_width_overflow) and [`min_wrap_width()`](https://docs.rs/html2text/latest/html2text/config/struct.Config.html#method.min_wrap_width).
